### PR TITLE
Feat/react container

### PR DESCRIPTION
**Pull request by literat**
Introducing React components for Container and Grid.

- [x] map prefix to classes after #110 is merged

### CODE REVIEW

File: packages/web-react/src/components/Container/Container.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import { applyClassNamePrefix } from '../../utils/classname';
+import { useClassNamePrefix } from '../../hooks/useClassNamePrefix';
+import { ChildrenProps, StyleProps } from '../../types';
+
+export interface ContainerProps extends ChildrenProps, StyleProps {
+  tag: ElementType;
+}
+
+const defaultProps = {
+  tag: 'div',
```
Comment (janicekt) [inline packages/web-react/src/components/Container/Container.tsx:12]: will it ever be anything other than div?

File: packages/web-react/src/components/Container/Container.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+};
+
+export const Container = ({ tag: Tag, className, children, ...restProps }: ContainerProps): JSX.Element => {
```
Comment (janicekt) [inline packages/web-react/src/components/Container/Container.tsx:15]: do we not count on a fluid container?

File: packages/web-react/src/components/Grid/Grid.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+};
+
+const Template = (args: SpiritGridProps) => <Grid {...args} />;
+
+export const ExampleGrid = Template.bind({});
+ExampleGrid.args = {
+  children: 'Grid',
+  mobile: 1,
+  tablet: 2,
+  desktop: 3,
```
Comment (janicekt) [inline packages/web-react/src/components/Grid/Grid.stories.tsx:26]: do you think this scale (mobile, table, desktop) will be enough? how do we want to scale this case?

File: packages/web-react/src/components/Container/Container.tsx (hunk 1 above)
Comment (literat) [reply packages/web-react/src/components/Container/Container.tsx:12]: Do not know. But it should be also `section` or another react component, maybe.

File: packages/web-react/src/components/Container/Container.tsx (hunk 2 above)
Comment (literat) [reply packages/web-react/src/components/Container/Container.tsx:15]: There is no `fluid` property now (https://github.com/lmc-eu/spirit-design-system/blob/main/packages/web/src/components/Container/_Container.scss). This is a question for @adamkudrna or @crishpeen, I think.

File: packages/web-react/src/components/Grid/Grid.stories.tsx (hunk 3 above)
Comment (literat) [reply packages/web-react/src/components/Grid/Grid.stories.tsx:26]: I think this scale is based on products needs and it is based on design/UX knowledge. I think that we can still go up to large or extra large devices.

**Comment by literat** [comment]
- [ ] Add missing props for `Grid--cols-12` class variant

File: packages/web-react/src/components/Container/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Container
+
+Container docs.
```
Comment (adamkudrna) [inline packages/web-react/src/components/Container/README.md:3]: I think we should start adding a meaning to these `README`s, this is not very helpful 🙂. Possible options:

- add some helpful content like example usage, link to Storybook, and/or props table,
- remove all React component `README`s altogether.

Either way, it should not be a pain for us—Supernova is on it's way so we should ideally put here something we will be able to reuse further down the road.

File: packages/web-react/src/components/Grid/Grid.tsx [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import React, { ElementType } from 'react';
+import classNames from 'classnames';
+import { useGridStyleProps } from './useGridStyleProps';
+import { SpiritGridProps } from '../../types';
+
+export const Grid = <T extends ElementType = 'div'>(props: SpiritGridProps<T>): JSX.Element => {
+  const { className, elementType: ElementType = 'div', children, ...restProps } = props;
```
Comment (adamkudrna) [inline packages/web-react/src/components/Grid/Grid.tsx:7]: Apart from `Container`, custom element other than `div` for a `Grid` is a perfectly common use case 👍🏻. This typically includes lists: `ol`, `ul`, `dl`.

File: packages/web-react/src/components/Grid/Grid.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+const Template = (args: SpiritGridProps) => <Grid {...args} />;
+
+export const EqualColumns = Template.bind({});
+EqualColumns.args = {
+  children: (
+    <>
+      <div style={exampleGridStyle}></div>
+      <div style={exampleGridStyle}></div>
+      <div style={exampleGridStyle}></div>
```
Comment (adamkudrna) [inline packages/web-react/src/components/Grid/Grid.stories.tsx:37]: I'm happy all these inline styles are just an example and they don't make it to production 😄.

File: packages/web-react/src/components/index.tsx [hunk 7]
```diff
@@ -1,2 +1,4 @@
 export * from './Button';
 export * from './Tag';
+export * from './Container';
```
Comment (adamkudrna) [inline packages/web-react/src/components/index.tsx:3]: Should come before `Tag`.

File: packages/web-react/src/types/grid.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { ElementType, JSXElementConstructor } from 'react';
+import { ChildrenProps, StyleProps } from './shared';
+
+export enum GridVariant {
+  XS = 1,
+  SM = 2,
+  MD = 3,
+  LG = 4,
+  XL = 6,
+  XXL = 12,
```
Comment (adamkudrna) [inline packages/web-react/src/types/grid.ts:10]: As discussed on Slack, please list just the numbers here. Introducing variant aliases is something that should be coined into our DS vocabulary so we would have to discuss with designers too. We can change it if there is a confusion on consumers' side. Is it OK?

File: packages/web-react/src/types/grid.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+export interface GridProps<T extends ElementType = 'div'> extends GridElementTypeProps<T> {}
+
+export interface SpiritGridProps<T extends ElementType = 'div'> extends GridProps<T>, ChildrenProps, StyleProps {
+  mobile: GridVariant;
+  tablet: GridVariant;
+  desktop: GridVariant;
+  narrow?: boolean;
+}
```
Comment (adamkudrna) [inline packages/web-react/src/types/grid.ts:28]: 👍🏻

File: packages/web-react/src/components/Grid/useGridStyleProps.tsx [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import { useClassNamePrefix } from '../../hooks/useClassNamePrefix';
+
+export interface GridStyles<T> {
+  /** className props */
+  classProps: string;
+  /** Props for the grid element. */
+  props: T;
+}
+
+const gridBreakpoints = ['mobile', 'tablet', 'desktop'];
```
Comment (adamkudrna) [inline packages/web-react/src/components/Grid/useGridStyleProps.tsx:13]: This is not just about the `Grid`, it's common breakpoints of our DS. Do you think we could store the constant somewhere higher in the source?

File: packages/web-react/src/components/Grid/__tests__/useGridStyleProps.test.tsx [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    expect(result.current.classProps).toBe('Grid');
+  });
+
+  it('should return narrow class', function () {
+    let props = { narrow: true };
+    let { result } = renderHook(() => useGridStyleProps(props));
+
+    expect(result.current.classProps).toBe('Grid Grid--narrow');
+  });
+  it('should return responzive variants', function () {
```
Comment (adamkudrna) [inline packages/web-react/src/components/Grid/__tests__/useGridStyleProps.test.tsx:20]: `responsive`

**Review by adamkudrna — CHANGES_REQUESTED** [review]
It's on a good way! A few things to point out:

- Please review the `Grid` variants enum as discussed on Slack.
- Please consider the `className` openness to public. May deserve a separate issue, I don't mean it to block it here.
- Please take a look at `README`s in `web-react`—they should say something, or they should not be there. I know there are already two empty `README`s, so feel free to get back to it later in a separate issue/PR.
- Please split the changes into two PRs next time (`Container` and `Grid`), less code is easier to review 🙏🏻.

Thank you!

File: packages/web-react/src/components/Container/Container.tsx [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-react/src/components/Container/Container.tsx:14]: I think that we should not be so strict. But anyway maybe we should place the className prop before our classes to be sure that our classes are prioritized. I know that there are ways to override this but when prohibiting `className` so we must also prohibit the rest of the props. When using `...restProps`, API is open anyway, you can use `style` for example.

So, in this component we can close it entirely or open it entirely. I do not want to do any props filtering now.

File: packages/web-react/src/components/index.tsx (hunk 7 above)
Comment (literat) [reply packages/web-react/src/components/index.tsx:2]: Done.

File: packages/web-react/src/types/grid.ts (hunk 8 above)
Comment (literat) [reply packages/web-react/src/types/grid.ts]: I have already changed it.

File: .storybook/assets/stylesheets/index.scss [hunk 13]
```diff
@@ -2,3 +2,5 @@
 
 @use '../../../packages/web/src/components/Button';
 @use '../../../packages/web/src/components/Tag';
+@use '../../../packages/web/src/components/Container';
```
[the remark this answers is not in the document]
Comment (literat) [reply .storybook/assets/stylesheets/index.scss:4]: Fixed.

File: packages/web-react/scripts/entryPoints.js [hunk 14]
```diff
@@ -5,6 +5,8 @@ const entryPoints = [
   { dirs: ['components'] },
   { dirs: ['components', 'Button'] },
   { dirs: ['components', 'Tag'] },
+  { dirs: ['components', 'Container'] },
```
[the remark this answers is not in the document]
Comment (literat) [reply packages/web-react/scripts/entryPoints.js:7]: Fixed.

File: packages/web-react/src/components/Grid/__tests__/useGridStyleProps.test.tsx (hunk 11 above)
Comment (literat) [reply packages/web-react/src/components/Grid/__tests__/useGridStyleProps.test.tsx]: Fixed.

File: packages/web-react/src/components/Container/Container.tsx (hunk 1 above)
Comment (literat) [reply packages/web-react/src/components/Container/Container.tsx]: This option I have removed and there is only `div` now.

File: packages/web-react/src/components/Grid/useGridStyleProps.tsx (hunk 10 above)
Comment (literat) [reply packages/web-react/src/components/Grid/useGridStyleProps.tsx:13]: Fixed in https://github.com/lmc-eu/spirit-design-system/pull/168

**Comment by literat** [comment]
The grid has been moved to separate PR: https://github.com/lmc-eu/spirit-design-system/pull/168 This was growing too big. Thank you @adamkudrna for the notice.

File: packages/web-react/src/components/Container/README.md (hunk 4 above)
Comment (literat) [reply packages/web-react/src/components/Container/README.md]: I would like to start adding links to the Storybook but I am waiting for Supernova. Storybook is now for me a visual testing tool, as the examples. However I have added basic usage information. So it has meaning now. I want to do it better in near future.

**Review by adamkudrna — APPROVED** [review]

File: packages/web-react/src/components/Container/Container.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (adamkudrna) [reply packages/web-react/src/components/Container/Container.tsx]: > place the className prop before our classes to be sure that our classes are prioritized

That's not going to work, class names order _in HTML_ doesn't play any role in CSS cascade.

> so we must also prohibit the rest of the props

Doesn't have to be so, it's not a black-and-white question (can I say that at all…? 🙊). Unwanted items [can be filtered out](https://github.com/react-ui-org/react-ui/blob/master/src/lib/components/_helpers/transferProps.js) from `...restProps`.

> I do not want to do any props filtering now.

OK, I see and understand.

**Review by adamkudrna — APPROVED** [review]
Cool, thank you! 👍🏻

Also thanks for splitting the PR, although I was far from asking to do it now 🙂. But it's definitely better now.

**Pull request merged**